be/src/exec/parquet_builder.h

```diff
@@ -0,0 +1,115 @@
+#include <string>
+
+#include "common/status.h"
+#include "exec/file_builder.h"
+#include "gen_cpp/DataSinks_types.h"
+#include "gen_cpp/PlanNodes_types.h"
+#include "gen_cpp/Types_types.h"
+#include "gen_cpp/parquet_types.h"
+
+namespace starrocks {
+
+class ExprContext;
+class FileWriter;
+
+struct ParquetBuilderOptions {
+    TCompressionType::type compression_type;
+    bool use_dictionary;
+    int64_t row_group_max_size;
+    std::vector<std::string> file_column_names;
+};
+
+class ParquetOutputStream : public arrow::io::OutputStream {
+public:
+    ParquetOutputStream(std::unique_ptr<WritableFile> _writable_file);
+    ~ParquetOutputStream() override;
+    arrow::Status Write(const void* data, int64_t nbytes) override;
+    arrow::Status Write(const std::shared_ptr<arrow::Buffer>& data) override;
+    arrow::Result<int64_t> Tell() const override;
+    arrow::Status Close() override;
+
+    bool closed() const override { return _is_closed; }
+    size_t get_written_len() const;
+
+private:
+    std::unique_ptr<WritableFile> _writable_file;
+    size_t _cur_pos = 0;
+    bool _is_closed = false;
+};
+
+class ParquetBuildHelper {
+public:
+    static void build_file_data_type(parquet::Type::type& parquet_data_type, const LogicalType& column_data_type);
+
+    static void build_parquet_repetition_type(parquet::Repetition::type& parquet_repetition_type,
+                                              const bool is_nullable);
+
+    static void build_compression_type(parquet::WriterProperties::Builder& builder,
+                                       const TCompressionType::type& compression_type);
+};
+
+class ParquetBuilder : public FileBuilder {
+public:
+    ParquetBuilder(std::unique_ptr<WritableFile> writable_file, const std::vector<ExprContext*>& output_expr_ctxs,
+                   const ParquetBuilderOptions& options);
+
+    ~ParquetBuilder() override = default;
+
+    Status add_chunk(Chunk* chunk) override;
+
+    std::size_t file_size() override;
+
+    Status finish() override;
+
+private:
+    Status _init(const ParquetBuilderOptions& options);
+    void _init_properties(const ParquetBuilderOptions& options);
+    Status _init_schema(const std::vector<std::string>& file_column_names);
+    void _generate_rg_writer();
+    void _rg_writer_close();
+    size_t _get_rg_written_bytes();
+    void _check_size();
+
+    std::unique_ptr<WritableFile> _writable_file;
+    std::shared_ptr<ParquetOutputStream> _output_stream;
+    const std::vector<ExprContext*>& _output_expr_ctxs;
+    std::shared_ptr<::parquet::schema::GroupNode> _schema;
+    std::shared_ptr<::parquet::WriterProperties> _properties;
+    std::unique_ptr<::parquet::ParquetFileWriter> _file_writer;
+    ::parquet::RowGroupWriter* _rg_writer = nullptr;
+    std::vector<int64_t> _buffered_values_estimate;
+    int64_t _row_group_max_size;
+    int64_t _cur_written_rows{0};
+    std::atomic<bool> _rg_writer_closing = false;
+    std::atomic<bool> _closed = false;
```

Comment:
Seems no need to record this atomic value, because file builder is call by single thread.